### PR TITLE
Lawn Mower Body

### LINKED ISSUE #331: Lawnmover body
A new body that reloads by shooting with a stricter and faster timing each consecutive firing. (Resets to slowest firing speed if timing is missed)

### PR DESCRIPTION
**Pull request by Fueredoriku**
Resolves #331 

- Add prefab, stats and animation
- Custom reload behaviour
- Screen shader material
- Lawn mower string is pulled by players at Fire
- Smoke VFX at success and another one for fail
mower

### CODE REVIEW

**Review by toberge — CHANGES_REQUESTED** [review]
image
Uh oh

**Review by toberge — CHANGES_REQUESTED** [review]
Tested with the cannon barrel and it is impossible to fire after the first timing, probably since this body doesn't account for reductions in fire rate...

File: Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    private void Fire(GunStats stats)
+    {
+        LeanTween.cancel(gameObject);
+        animator.SetTrigger("PistonPump");
+        handAnimator.SetTrigger("Pull");
+        exhaustParticles.Play();
+        var currentDegrees = mowerScreen.GetFloat("_ArrowDegrees");
+        if (currentDegrees > currentTarget - arrowHitSpan && currentDegrees < currentTarget + arrowHitSpan)
```
Comment (toberge) [inline Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs:77]: Perhaps you could extract these conditions into variables so the if check becomes a little easier to read?

File: Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        gameObject.LeanValue( 
+            (degree) => mowerScreen.SetFloat("_ArrowDegrees", degree), minArrowDegrees, maxArrowDegrees, 3f)
+            .setOnComplete(() => 
+            {
+                currentTarget = maxArrowDegrees;
+                mowerScreen.SetFloat("_TargetDegrees", currentTarget);
+                Reload(stats);
+                gameObject.LeanValue(
+                    (degree) => mowerScreen.SetFloat("_ArrowDegrees", degree), maxArrowDegrees + 4f, maxArrowDegrees - 4f, 0.15f)
+                    .setLoopPingPong();
```
Comment (toberge) [inline Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs:108]: This is (almost) the same code as in the Start() method, perhaps you could extract this to its own method? e.g. ResetArrow or ResetProgression?

**Review by toberge — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    private ParticleSystem exhaustParticles;
+    [SerializeField]
+    private ParticleSystem failedExhaustParticles;
+    [SerializeField]
+    private PlayerHand playerHand;
+    [SerializeField]
+    private LineRenderer handString;
+    private Animator handAnimator;
+    [SerializeField]
+    private Transform lineStart;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs]: Done

File: Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            Debug.Log("MowerBody not attached to gun parent!");
+            return;
+        }
+        gunController.onFire += Fire;
+
+        currentTarget = maxArrowDegrees;
+        mowerScreen.SetFloat("_TargetDegrees", currentTarget);
+        mowerScreen.SetFloat("_ArrowDegrees", maxArrowDegrees);
+        gameObject.LeanValue(
+            (degree) => mowerScreen.SetFloat("_ArrowDegrees", degree), maxArrowDegrees + 4f, maxArrowDegrees - 4f, 0.15f)
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs]: This is not the tween for the arrow from min to max, but rather the wobbling arrow when arrow is at max. I have moved this to a named variable so it's more readable.

File: Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs (hunk 1 above)
Comment (Fueredoriku) [reply Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs]: Done

File: Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs (hunk 2 above)
Comment (Fueredoriku) [reply Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs]: Done 👍

Comment (toberge) [reply Assets/Scripts/Augment/AugmentImplementations/LawnMower.cs]: Well... I meant to include that entire lambda body in "the same code as in Start", and I'd imagined the refactor as passing a ResetArrow method as the LeanTween callback and calling said method in Start. Just do as you wish with my _actual_ suggestion.

**Review by toberge — APPROVED** [review]
Works with the cannon barrel and the arm is gone from the item pedestal, nice :+1: 

I'm still not entirely satisfied with some variable names and so forth, but I won't pester you too much about that. I'm also curious to see how this body fares in future playtests, as I fear it might require more skill than it is worth.

**Pull request merged**